# Gauss Elimination and Back Substitution in Python 
## Solving a system of linear algebraic equations 
$$\textbf{A}x=\textbf{b}$$
Example: Consider the following system of linear algebraic equations
$$x_1 - 3x_2 + x_3 = 4$$
$$2x_1 - 8x_2 + 8x_3 = -2$$
$$-6x_1 + 3x_2 - 15x_3 = 9$$

In [245]:
import numpy as np

The coefficient matrix from the above system is 
$$\begin{bmatrix} 
1 & -3 & 1\\
2 & -8 & 8\\
-6 & 3 & -15
\end{bmatrix}$$
and the b vector 
$$
\begin{bmatrix}
4\\-2\\9
\end{bmatrix}
$$

In [246]:
coefficient_matrix = np.array([[1,-3,1],
                              [2,-8, 8],
                              [-6,3,-15]])
b_vector = np.array([[4],
                    [-2],
                    [9]])
num_unknowns = coefficient_matrix.shape[1]
solution_vector = np.zeros(num_unknowns)

## Generate augmented matrix from the coefficient matrix and b vector
Together, the coefficient matrix and b vector form the augmented matrix
$$\begin{bmatrix} 
1 & -3 & 1 & | & 4\\
2 & -8 & 8 & | & -2\\
-6 & 3 & -15 & | & 9
\end{bmatrix}$$


In [247]:
augmented_matrix = np.hstack((coefficient_matrix,b_vector))

In [248]:
print("Augmented matrix:\n {}".format(augmented_matrix))
print("Number of unknowns: {}".format(num_unknowns))

Augmented matrix:
 [[  1  -3   1   4]
 [  2  -8   8  -2]
 [ -6   3 -15   9]]
Number of unknowns: 3


## Perform Gauss Elimination
The goal of gauss elimination is to form an upper trianguler matrix through elementary row operations on the augmented matrix. After this is achieved, we can use back substitution to solve for the 3 unknowns $x_1, x_2$ and $x_3$

In [249]:
for i in np.arange(num_unknowns):
    for j in np.arange(i+1, num_unknowns):
        factor = augmented_matrix[j][i]/augmented_matrix[i][i]
        for k in np.arange(num_unknowns+1):
            augmented_matrix[j][k]=augmented_matrix[j][k]- factor*augmented_matrix[i][k]
print(augmented_matrix)

[[  1  -3   1   4]
 [  0  -2   6 -10]
 [  0   0 -54 108]]


The augmented matrix is now the upper triangular matrix
$$
\begin{bmatrix}
1 & -3 & -1 & | & 4\\
0 & -2 & 6 & | & -10\\
0 & 0 & -54 & | & 108\\
\end{bmatrix}
$$

## Perform back-substitution to find solution vector

In [250]:
solution_vector[num_unknowns-1] = augmented_matrix[num_unknowns-1][num_unknowns]/(augmented_matrix[num_unknowns-1][num_unknowns-1])
for i in np.arange(num_unknowns-2,-1,-1):
    solution_vector[i]=augmented_matrix[i][num_unknowns]
    for j in np.arange(i+1,num_unknowns):
        solution_vector[i]-=augmented_matrix[i][j]*solution_vector[j]
    solution_vector[i]=solution_vector[i]/augmented_matrix[i][i]

print("Solution:")
for i in np.arange(len(solution_vector)):
    print("x_{} = {}".format(i+1, solution_vector[i]))

Solution:
x_1 = 3.0
x_2 = -1.0
x_3 = -2.0


## Solution 
$$3 - 3(-1) + -2 = 4$$
$$2(3) - 8(-1) + 8(-2) = -2$$
$$-6(3) + 3(-1) - 15(-2) = 9$$